sdk/identity/Azure.Identity/src/TenantIdResolver.cs

```diff
@@ -0,0 +1,26 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using Azure.Core;
+
+namespace Azure.Identity
+{
+    internal static class TenantIdResolver
+    {
+        /// <summary>
+        /// Resolves the tenantId based on the supplied configuration values.
+        /// </summary>
+        /// <param name="explicitTenantId">The tenantId passed to the ctor of the Credential.</param>
+        /// <param name="context">The <see cref="TokenRequestContext"/>.</param>
+        /// <param name="options">The <see cref="TokenCredentialOptions"/>.</param>
+        /// <returns></returns>
+        public static string Resolve(string explicitTenantId, TokenRequestContext context, TokenCredentialOptions options)
+        {
+            return options?.AllowMultiTenantAuthentication switch
+            {
+                false => explicitTenantId ?? context.TenantId,
+                _ => context.TenantId ?? explicitTenantId
+            };
```

Comment:
I know it's somewhat of a breaking change but I think that this should be simplified to throw if `AllowMultiTenantAuthentication` is false and the `context.TenantId` doesn't match `explicitTenantId`. I think this is more the behavior customers would expect if they are not allowing multi-tenant auth. Simply returning a token from the tenant specified on credential construction rather than the requested one almost seems like a silent failure to me.  It would be good to get @KrzysztofCwalina and @tg-msft thoughts on whether this "breaking change" would be acceptable.

Reply:
A couple of questions:
- What is the default value of `AllowMultiTenantAuthentication`?
- Is `context.TenantId` that property that Chris just added (and nobody could possibly be using yet)?
- How could this go wrong in practice?  I'm calling a service that supports discovery, I turn discovery off/give a specific TenantId, but then we get a challenge and do whatever dance to get a different tenant that doesn't match?

Reply:
> * What is the default value of `AllowMultiTenantAuthentication`?

The default is false

> * Is `context.TenantId` that property that Chris just added (and nobody could possibly be using yet)?

Yes

> * How could this go wrong in practice?  I'm calling a service that supports discovery, I turn discovery off/give a specific TenantId, but then we get a challenge and do whatever dance to get a different tenant that doesn't match?

I believe the concern is that we could send a token request to an authority that wasn't intended.